Chapter5 (#51)
UBC-MDS/ds-toolbox: chapters/en/module5.md

</opt>
<opt text='<code>HEAD -> app</code> ' correct='true'>
'<code>HEAD</code> is pointing to the last place you make changes and that was in the '<code>app</code> branch
</opt>
<opt text='<code>main</code>'>
In what branch are you working now?
</opt>
<opt text=' <code>HEAD -> main</code>'>
In what branch are you working now? 
</opt>

</choice>

</exercise>

<exercise id='3' title="What is the relationship between the branches?"  type='slides, video'>
<slides source='module5/module5_02' shot='0' start='3:42' end='4:35'> </slides>
</exercise>

<exercise id='4' title='Observing the trees'>

**True or False**

The labels <code>origin/main</code> and <code>origin/HEAD</code> always will be associated to the more recent commit in the history of a project that you have just cloned from GitHub.

<choice id = 4>
<opt text='True'>
What happens if the last change in the GitHub repository was in a different branch than <code>main</code>? You can try to reproduce this in your computer, creating a new branch in the remote and pulling the information to the local repository!
</opt>
<opt text='False'  correct='true' >

> The term `origin` refers to the remote repository. If the last change in the remote repository you are cloning was in other branch than <code>main</code> the label in the most recent commit will be (<code>origin/new-branch</code>). <code>origin/HEAD</code> will be always pointing the last commit in the <code>main</code> branch next to <code>origin/main</code>.

Author: I get your point but for the purposes of the exercise the following change potentially could be better:

> The term `origin` refers to the remote repository. If the last change in the remote repository you are cloning was in other branch than <code>main</code> the label in the most recent commit will be (<code>origin/new-branch</code>). <code>origin/HEAD</code> will be always pointing the last commit in the <code>main</code> branch next to <code>origin/main</code> (except you change the GitHub settings).

What do you think?

> The term `origin` refers to the remote repository. If the last change in the remote repository you are cloning was in other branch than <code>main</code> the label in the most recent commit will be (<code>origin/new-branch</code>). <code>origin/HEAD</code> will be always pointing the last commit in the <code>main</code> branch next to <code>origin/main</code>.

Reviewer: Sounds good! Just some small word changes:

> The term `origin` refers to the remote repository. If the last change in the remote repository you are cloning was in another branch than <code>main</code> the label in the most recent commit will be (<code>origin/new-branch</code>). <code>origin/HEAD</code> will always be pointing to the last commit in the <code>main</code> branch next to <code>origin/main</code> (except if you change the default branch setting in git).

</opt>
</choice>

After running <code> git log --oneline --all --graph </code> in your repository you get as output the following tree:

* a949e61 2020-09-13 (HEAD -> master) Update readme
| * 0a8dfef 2020-09-13 (lecture-todo) List todo items
|/
* 6e69947 2020-09-13 (origin/master, origin/HEAD) Add license

> * 0d5a6f7 2020-09-13 Initial commit

Reviewer: Currently it actually shows up with the wrong formatting

![image](https://user-images.githubusercontent.com/4560057/178121042-e39fe7f9-9dda-4bcf-8061-b7603b778bf8.png)

Select the options that are TRUE

<choice id = 5>
<opt text='The most recent commit on GitHub was performed on the <code>lecture-todo</code> branch'>
You can recognize the latest commit on GitHub with the label <code> origin/HEAD </code>.
</opt>

> <opt text='There is a branch called <code> lecture-todo </code> that has not been pushed to GitHub yet ' correct='true'>

Reviewer: ```suggestion
<opt text='There are changes to the <code> lecture-todo </code> branch that has not been pushed to GitHub yet ' correct='true'>
```

Technically the branch could exist but be older than the three commits we see here

> The labels <code> origin/master</code> , origin/HEAD</code> that points the last change in the remote are downstream. This means that a949e61 and 0a8dfef were created after pulling or cloning from GitHub.

Reviewer: ```suggestion
The <code> lecture-todo </code> branch shows up alone on the latest commit. If the changes were on GitHub there would be at least one branch with the <code> origin/ </code> prefix on the latest commit as well.
```

</opt>
<opt text='The most recent local commit of this repository has associated the message "Update readme"' >

> The most recent commits are always the first to appear in the representation of the trees. 

Reviewer: ```suggestion
By default, the most recent commits appear on the top of commit history. 
```

</opt>
<opt text='The commits a949e61 and 0a8dfef exist only in your local computer' correct='true'>
The labels <code> origin/master</code>, <code>origin/HEAD</code> that are pointing the last change in the remote are downstream these two commits. This means that a949e61 and 0a8dfef were created after pulling or cloning from GitHub.
</opt>
<opt text='The commits 6e69947 and 0d5a6f7 exist both in your local computer and on GitHub' correct='true'>

> The labels <code> origin/master</code>, origin/HEAD</code> that points the last change in the remote are upstream these two commits. This means that 6e69947 and 0d5a6f7 were pulled or cloned from GitHub.

Reviewer: ```suggestion
The position of the label <code> origin/master</code> indicates that both these commits are part of this branch on the GitHub remote.
```

</opt>
<opt text='the <code>app</code> branch with the <code>main</code> branch'>
How do you know you are in the <code>main</code> branch? Where you can find which branch you are currently working on?
</opt>
<opt text='the <code>fix-app-slider</code> branch with the <code>main</code> branch'>
How do you know you are in the <code>main</code> branch? Where you can find which branch you are currently working on?
</opt>
<opt text='the <code>canadian-languages</code> branch with the <code>app</code> branch'>
Is <code> canadian-languages</code>  referring to a branch? 
</opt>
</choice>
</exercise>

<exercise id='7' title='When merging branches goes social: opening pull requests'  type='slides, video'>
<slides source='module5/module5_04' shot='0' start='3:42' end='4:35'> </slides>
</exercise>

<exercise id='8' title='You are cordially invited to review my changes'>

**True or false**

You should be quite sure when creting a pull request because you will not be able add more commits after you create it.

<choice id = 8>
<opt text='True' >
Try open a pull request and pushing new commits and check what happen!
</opt>
<opt text='False' correct='true'>

> You can continue adding commits to a pull request pushing to the branch you used to create it. You will see the commits as a timeline when you open it on GitHub. Try it!

Reviewer: ```suggestion
You can continue adding commits to a pull request by pushing to the branch you used to create it. You will see the commits as a timeline when you open it on GitHub. Try it!
```

</opt>
</choice>

**True or false**  

It is possible to open a pull request locally

<choice id = 9>
<opt text='True' >

> It is possible to <a href="https://docs.github.com/es/enterprise-server@3.1/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally">check out a pull request locally to solve a merge conflict or test changes before aproving it</a>, but not actually create one.

Reviewer: ```suggestion
It is possible to <a href="https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally">check out a pull request locally to solve a merge conflict or test changes before approving it</a>. If you wanted to created a pull request locally, you would need to use <a href="https://cli.github.com/"> the GitHub CLI</a>.
```

</opt>
<opt text='False' correct='true'>
You have to enter to your GitHub repository to start a pull request. What you can do locally is to create a new branch and push it to GitHub. Then GitHub will give you the option to start a pull request.
</opt>
</choice>

> How is it different to open a pull request to merging branches? Is it not basically the same?

Reviewer: ```suggestion
Is opening a pull request essentially the same as merging a branch locally?
```

Select the correct option

<choice id = 10>

> <opt text='Yes, it is the same, you are merging branches but using GitHub' >

Reviewer: ```suggestion
<opt text='Yes, it is the same, you are just merging branches, either using the GitHub UI or your local interface' >
```

> Even if the final goal of the pull request is to merge your changes with other branch, it is also a way to socialize that you created those changes and also to involve other team members, for example, as reviewers.

Reviewer: ```suggestion
Even if the final goal of the pull request is to merge your changes with another branch, it is also a way to communicate that you created those changes and to involve other team members, for example as reviewers.
```

</opt>
<opt text='You open a pull request to merge branches when you want to involve your collaborators' correct='true'>
</opt>
</choice>
</exercise>

<exercise id='9' title='About the different ways we can collaborate' type='slides, video'>
<slides source='module5/module5_05' shot='0' start='3:42' end='4:35'> </slides>
</exercise>

<exercise id='10' title='Between branches and forks'>

> In which cases branching would be better that using forks?

Reviewer: ```suggestion
In which cases would branching be better than using forks?
```

<choice id = 11>
<opt text='The owner of the repository have set branch protection rules to the <code>main</code>  branch'>

> Even if it is a good practice as avoid that contributors push code directly to the <code>main</code> branch by accident, protecting the branch it is not a requirement to collaborate using branching. 

Reviewer: ```suggestion
Even if it is a good practice to avoid that contributors push code directly to the <code>main</code> branch by accident, protecting the branch it is not a requirement to collaborate using branching. 
```

If you have not write access it would be impossible to contribute without forking the repository first.
</opt>
</choice>

What of these names are good for branches?

<choice id = 12>
<opt text='arman, florencia, joel'>
Using the names of collaborators as names of branches is not meaningful as it is not providing information about what each member of the team is working on.
</opt>
<opt text='fix-dockerfile, feature-testing, feature-app-modules' correct='true'>
There is not a unique way to do this, you will have to define with your team a rule or follow the criteria of the repository you are contributing.
</opt>
<opt text='1616789, 1235348, 1555684' >
Using only numbers could create confusion when merging the branches. For example, selecting as branch name the issue id that the branch is trying to solve will seem a logical choice, but it is uninformative and could lead to mistakes.
</opt>
</choice>
</exercise>

<exercise id='11' title='GitHub cutlery: introducing the forks' type='slides, video'>
<slides source='module5/module5_06' shot='0' start='3:42' end='4:35'> </slides>
</exercise>

<exercise id='12' title='Forking your repository'>

**True or false**

> When working on a forked repository, it is a good practice to begin a new branch before submitting a pull request.

Reviewer: ```suggestion
When working on a forked repository, it is good practice to begin a new branch before submitting a pull request.
```

<choice id = 13>
<opt text='True' correct='true' >

> Making the pull request from a branch with a meaningful name will give the pull request's receivers useful information. If not they will receive changes of <code>main</code> branches from different users.
> </opt>
> <opt text='False'>
> </opt>

Reviewer: ```suggestion
</opt>
<opt text='False'>
Making the pull request from a branch with a meaningful name will give the pull request's reviewers useful information. If not they will receive changes of <code>main</code> branches from different users.
</opt>
```

</choice>

**True or false**

After forking a GitHub repository, you should clone the repository's copy in your GitHub account to begin working on it locally. 

<choice id = 14>
<opt text='True'  correct='true'>
You probably don't have write access to the original repository to push modifications directly to it, that is why you have to clone your fork.
</opt>
<opt text='False'>

> You will clone, pull, and push to your GitHub account's cloned repository. Contributions to the original GitHub repository will be made by pul requests from your fork to the original repository. 

Reviewer: ```suggestion
You will clone, pull, and push to your GitHub account's cloned repository (the fork). Contributions to the original GitHub repository will be made by pul requests from your fork to the original repository. 
```
